[WIP] adding support for multisite on cephadm

rkachach opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

phlogistonjohn commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,25 +102,75 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
+                err = f'{e.message}'
+                break
+
+        return HandleCommandResult(retval=retval, stdout=f'{outs}', stderr=f'{err}')
```
If I understand the code correctly, `outs` is a list of strings and thus the stdout here will be a repr of the list (like: `["foo", "ar"]`)
I suggest using JSON formatting instead of relying on a python repr. Especially since this is a  net-new command there's no backwards compatibility to worry about so we should start off with standardized structured output. JSON is more standard then python repr.   (It would also make a future conversion to the new decorators I'm working on easier. hint hint :-) )

rkachach replied:
TBH I wasn't sure which output format to use (so I used python repr) but your arguments are solid enough to change it to JSON. I'll be fixing it :+1:

rkachach commented:
jenkins retest this please

adk3798 commented on doc/mgr/rgw.rst:
```diff
@@ -37,6 +37,25 @@ system user has no more access keys, it is removed.
+Create a new realm + zonegroup + zone and deploy rgw daemons via the
+orchestrator using the information specified in the YAML file. In case
+the realm, zonegroup or zone already exist the module will not re-create
+them again, instead it will use their information. Following is an example
```
"instead it will use their information" for what purpose? I think if we're going to mention that the info from the existent realm/zonegroup/zone is getting used somewhere else we should also say for what.

rkachach replied:
text has changed

adk3798 commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,25 +103,78 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
                                  uid: Optional[str] = None,
-                                 start_radosgw: Optional[bool] = True):
+                                 start_radosgw: Optional[bool] = True,
+                                 zonegroup_is_master: Optional[bool] = True,
+                                 zone_is_master: Optional[bool] = True):
```
are we going to have any docs on what these two `is_master`  or the `start_radosgw` settings do?

rkachach replied:
code has changed we do not provide the user the possibility to configured master (we use master by default).

adk3798 commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,25 +103,78 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
+                    err = {'error': 'spec fields: realm_name, zonegroup_name and zone_name are mandatory',
+                           'retcode': errno.EINVAL}
+                    return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr=json.dumps(err))
+                retval, out, err_msg = RGWAM(self.env).realm_bootstrap(realm_name,
```
I feel like it would be worth having the option to specify a placement in the spec here if you want the orchestrator to deploy the rgw daemons. Otherwise, we're just going to place them on random nodes and users will have to go manually edit the spec the orchestrator has to say where they actually want them.

rkachach replied:
New code uses cephadm RGWSpec directly this way we can use the same params already supported by cephadm when specifying RGW params (including placement).

cbodley commented:
cc @yehudasa 

revisiting the description from https://github.com/ceph/ceph/pull/42710, i think we need to run the `ceph rgw realm bootstrap` part only once to create the original realm/zonegroup/zone

additional zones need to use the `ceph rgw zone create` part, providing the same realm_token from bootstrap

this workflow is necessary to support deployments that span ceph clusters

rkachach commented:
jenkins test api

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
> cc @yehudasa
> 
> revisiting the description from #42710, i think we need to run the `ceph rgw realm bootstrap` part only once to create the original realm/zonegroup/zone
> 
> additional zones need to use the `ceph rgw zone create` part, providing the same realm_token from bootstrap
> 
> this workflow is necessary to support deployments that span ceph clusters

Great, thank you for the hint. I updated and fixed the code. It would be great if you can have a look to the new changes :+1:

alimaredia commented on doc/cephadm/services/rgw.rst:
```diff
@@ -83,16 +83,16 @@ To deploy RGWs serving the multisite *myorg* realm and the *us-east-1* zone on
 Note that in a multisite situation, cephadm only deploys the daemons.  It does not create
-or update the realm or zone configurations.  To create a new realm and zone, you need to do
-something like:
+or update the realm or zone configurations.  To create a new realm and zone, you can use
```
Maybe add only creates new realms, zones and zonegroups here

rkachach replied:
Fixed

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+                                                  uid_prefix='user-sys',
                                                   is_system=True)
+            sys_user = RGWUser(sys_user_info)
+            logging.info('Created system user: %s' % sys_user.uid)
```
NIT, can we include zonegoup and realm for the user in the log statement

rkachach replied:
Fixed

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
         try:
             user_info = self.user_op().create(zone, zg, uid=uid, is_system=False)
+            user = RGWUser(user_info)
+            logging.info('Created regular user: %s' % user.uid)
```
same NIT, can we include zonegoup and zone in this log?

rkachach replied:
Fixed

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+        if zone_name in self.zone_op().list():
+            raise RGWAMException(f'Zone {zone_name} already exists')
+
+        # Create RGW multisite entities and update the period
```
is there something being done here specific to multisite or can this bootstrapped realm be used for single site as well?

rkachach replied:
Normally the realm/zonegroup/zone creation is the same for both single and multisite configuration. I mean, AFAIK multi-site in fact only makes sense when you create a secondary zone on another cluster and you configure it to sync with the primary zone.

adk3798 replied:
okay, was just asking as the comment specified "multisite entitities". Sounds like they are general use though which is nice.

rkachach replied:
I removed the mention to multisite from the comment since this is not only limited to the multisite case.

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+
+        if start_radosgw:
+            # Instruct the orchestrator to start RGW daemons, asynchronically, this will
+            # call back the rgw module to update the master zone with the corresponding endpoints
```
going of the code in cephadm, we only update these endpoints if `rgw_spec.update_endpoints and rgw_spec.rgw_realm_token is not None`. Do we have those guarantees with the spec we have here (we might, legitimately asking)?

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +669,159 @@ def realm_rm_zone_creds(self, realm_token_b64):
-        access_key = realm_token['access_key']
-        secret = realm_token['secret']
+        if not realm_token_b64:
+            raise RGWAMException('missing realm access config')
```
is a realm token equivalent to an access config? Intuitively I wouldn't think so. Is this some rgw terminology I'm not aware of?

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +669,159 @@ def realm_rm_zone_creds(self, realm_token_b64):
+    def zonegroup_create(self, realm_token_b64, zonegroup_name=None,
+                         endpoints=None, zonegroup_is_master=True):
+        if not realm_token_b64:
+            print('missing realm access config')
```
was going to comment on the use of `print` here but I see it is used elsewhere in this file as well. Does it actually work properly? Typically in mgr module stuff we always have to log, raise an exception or return things.

rkachach replied:
fixed

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +669,159 @@ def realm_rm_zone_creds(self, realm_token_b64):
+        logging.info(f"Pulled realm {realm_info['name']} ({realm_info['id']})")
+        realm_name = realm_info['name']
+        realm_id = realm_info['id']
+        logging.info(f"Pulled realm {realm_name} ({realm_id})")
```
duplicate log statement as the one a few lines above?

rkachach replied:
fixed

adk3798 reviewed (commented):
Just a more general review as I'm not an expert on proper procedure for actually setting up multisite. Will need further review from rgw expert(s) as well.

alimaredia commented:
@rkachach This seems to look good to me. Are there any tests that will be added for this feature in teuthology?

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+
+  rgw realm bootstrap [--realm-name] [--zonegroup-name] [--zone-name] [--port] [--placement] [--start-radosgw]
+
+The command supports providing the configuration through a spec file (`-i option`):
```
I have no strong opinion in this case however I think it's good to have it as well (even when already stated in the previous sentence) so if somebody is not familiar with the spec files he can still understand how are they passed.

adk3798 replied:
sure

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
-Removal of this token would remove the credentials, and if the corresponding
-system user has no more access keys, it is removed.
 
+User can list the available tokens for the created (or already existing) realms.
```
fixed

rkachach commented:
> @rkachach This seems to look good to me. Are there any tests that will be added for this feature in teuthology? Every other ceph orchestrator (rook, ceph-ansible) has integration tests for their implementation of this feature.

Thanks @alimaredia for reviewing the PR. I'll be adding tests and other code quality related changes such as typing and style changes on a separeted PR. I just didn't want to include a lot of changes on this first PR to make it easy for reviewers so it only includes functional and new logic code related to the rgw multisite configuration.

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
-multisite sync operations, and returns a corresponding token. It sets
-up new RGW instances via the orchestrator.
+multisite sync operations. Under the hood this module instructs the
+orchestrator to create and deploy the corresponding RGW daemons. The module
```
@adk3798 I tested this case and right now the module create the realm/zonegroup/zone and then fails to create the daemons showing the following error:

```
[ceph: root@ceph-node-0 /]# ceph rgw realm bootstrap -i test.yaml 

Error EINVAL: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 1567, in _oremote
    return mgr.remote(o, meth, *args, **kwargs)
  File "/usr/share/ceph/mgr/mgr_module.py", line 2147, in remote
    args, kwargs)
ImportError: Module not found

```

I think it would be better to check if the orchestrator is available or not before proceeding with the realms creation. WDYT?

rkachach replied:
@adk3798 I added a check to make sure cephadm is available before running the cmds.

alimaredia commented:
To make someone elses from the RGW team's review easier I'm going to give my breakdown of the steps to run multisite and what I saw as the commands and code paths that accomplished those steps.
@rkachach  If I misinterpreted some of these ceph-adm/mgr module code paths let me know.

### Non-ceph-adm Radosgw-admin commands:
#### Primary Cluster (PC) Steps:
(PC.1) radosgw-admin realm create --rgw-realm=realm1
(PC.2) radosgw-admin zonegroup create --rgw-realm=realm1 --rgw-zonegroup=zg1 --master --endpoints=http://127.0.0.1:8001
(PC.3) radosgw-admin zone create --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone1 --master
(PC.4) radosgw-admin user create --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone1 --uid=testuser --display-name=testuser --access-key=akey --secret=skey --system
(PC.5) radosgw-admin zone modify --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone1 --endpoints=http://127.0.0.1:8001 --access-key=akey --secret-key=skey
(PC.6) radosgw-admin period update --commit --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone1
(PC.7) start the rgw on 127.0.0.1:8001 OR restart the rgw with the rgw realm, zg, zone configured for it

#### Secondary cluster (SC) Steps:
(SC.1) radosgw-admin realm pull --url=http://127.0.0.1:8001 --access-key=akey --secret=skey
(SC.2) radosgw-admin zone create --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone2 --access-key=akey --secret-key=skey
(SC.3) radosgw-admin period update --commit --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone2
(SC.4) # start the rgw

If necessary:
(SC.5) radosgw-admin zone modify --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone2 --endpoints=http://127.0.0.1:8002 --access-key=akey --secret-key=skey
(SC.3) radosgw-admin period update --commit --rgw-realm=realm1 --rgw-zonegroup=zg1 --rgw-zone=zone2

### Ceph-adm version:
Commands & code paths that do the equivalent of above assuming correct config spec of master zone, master zone group, endpoints:

#### Primary Cluster Steps:
`ceph rgw realm bootstrap`
  (PC.1) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> create_realm() -> RealmOp.create()
  (PC.2) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> create_zonegroup() -> ZonegroupOp.create()
  (PC.3) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> create_zone() -> ZoneOp.create()
  (PC.4) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> create_user() -> UserOp.create()
  (PC.5) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> zone_modify() -> ZoneOp.modify()
  (PC.6) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> update_period() -> PeriodOp.update()
  (PC.7) above command calls _cmd_rgw_realm_bootstrap -> realm_bootstrap -> apply_rgw()

`ceph rgw zone update` (with zone endpoints in spec)
  (PC.5) above command calls zone_modify() -> ZoneOp.modify()
  (PC.6) above command calls zone_modify() -> update_period() -> PeriodOp.update()

#### Secondary cluster (SC) Steps:
`ceph rgw zone create`
  (SC.1) above command calls _cmd_rgw_zone_create -> zone_create() -> RealmOp.pull()
  (SC.2) above command calls _cmd_rgw_zone_create -> zone_create() -> create_zone() -> ZoneOp.create()
  (SC.3) above command calls _cmd_rgw_zone_create -> zone_create() -> update_period() -> PeriodOp.update()
  (SC.4) above command calls _cmd_rgw_zone_create -> zone_create() -> apply_rgw()

`ceph rgw zone update` (with zone endpoints in spec)
  (SC.5) above command calls zone_modify() -> ZoneOp.modify()
  (SC.3) above command calls zone_modify() -> update_period() -> PeriodOp.update()

The steps in the `ceph rgw` commands map to the underlying `radosgw-admin` commands just fine. The PR that adds the feature to an integration testing suite should close any gaps, if any, for this feature.

alimaredia commented on doc/cephadm/services/rgw.rst:
```diff
@@ -83,16 +83,16 @@ To deploy RGWs serving the multisite *myorg* realm and the *us-east-1* zone on
 
 .. prompt:: bash #
 
   radosgw-admin realm create --rgw-realm=<realm-name> --default
```
I don't think you need `--default` for the below realm/zonegroup/zone create commands below. That flag only applies if the realm/zg/zone is the only one in the cluster/realm/zonegroup.

rkachach replied:
Fixed

rkachach replied:
fixed

alimaredia commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+    }
+  ]
+
+User can use the token to create and synchronize a secondary zones
```
I'd suggest something like: `User can use the token to pull a realm to create secondary zone on a different cluster that syncs with the primary ...`

rkachach replied:
fixed

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +191,84 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+        return HandleCommandResult(retval=0, stdout=json.dumps(realms_info), stderr='')
+
+    @CLICommand('rgw zone update', perm='rw')
+    def update_zone_info(self, realm_name: str, zonegroup_name: str, zone_name: str, realm_token: str, endpoints: List[str]):
```
where are the `endpoints` coming from here? Is it from the zone spec as documented in `mgr/rgw.rst`?

rkachach replied:
These endpoints are provided by cephadm automatically (and asynchronously). The call is made from `mgr/cephadm/server.py::_update_rgw_endpoints` which is called once the rgw daemons are created successfully by cephadm.

alimaredia commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,110 @@ RGW Realm Operations
+  rgw_realm: myrealm
+  rgw_zonegroup: myzonegroup
+  rgw_zone: my-secondary-zone
+  rgw_realm_token: <token>
```
@rkachach have you considered having the realm name and the zonegroup name in this token? This way when you pass it around you don't need to have the realm and zonegroup names in this spec. Just the zone name, the token, and the below information.

rkachach replied:
yeah, actually the realm and zonegroup are not used in this case. I modified the code the get rid of them since we only need the zone and the token in order to create the secondary zone.

rkachach replied:
I changed the code so the user only has to provide the realm token + zone. Realm name is derived automatically from the token.

alimaredia commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,110 @@ RGW Realm Operations
-  ceph rgw zone create
+  ceph rgw zone create -i spec.yaml
 
 Create a new zone and join existing realm (using the realm token)
```
This could be `Join an existing realm by creating a new secondary zone (using the realm token)"

rkachach replied:
fixed

alimaredia commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -814,6 +814,28 @@ def update_progress() -> None:
+                    for s in self.mgr.cache.get_daemons_by_service(rgw_spec.service_name()):
+                        if s.ports:
+                            for p in s.ports:
+                                ep.append(f'http://{s.hostname}:{p}')
```
Can these endpoints be added manually instead and only manually? We found that works best in rook. I can go into more detail why.

rkachach replied:
> Can these endpoints be added manually instead and only manually? We found that works best in rook. I can go into more detail why.

I updated the code to check if the spec has `ssl` so in this case we use https.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +191,84 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+                             placement: Optional[str] = None,
+                             start_radosgw: Optional[bool] = True,
+                             inbuf: Optional[str] = None):
         """Bootstrap new rgw zone that syncs with existing zone"""
```
```suggestion
        """Bootstrap new rgw zone that syncs with zone on another cluster in the same realm"""
```

rkachach replied:
Fixed

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +162,83 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+                return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr='Invalid arguments: -h or --help for usage')
+
+            for rgw_spec in rgw_specs:
+                retval, out, err = RGWAM(self.env).zone_create(rgw_spec, start_radosgw)
```
I'm having a hard time understand the conditions that would have more than 1 rgw_spec. Could I get an example?

rkachach replied:
@adk3798 the output is informative only. I fixed it to show all the created zones.
@alimaredia this support is something we normally provide on cephadm also so user can use the same yaml to create several entries (instead of having to run the cmd for each spec). For example, imagine the case when a user wants to create several secondary zones so he can provide a single Yaml spec that defines all of them.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +191,84 @@ def _cmd_rgw_realm_rm_zone_creds(self,
         try:
-            retval, out, err = RGWAM(self.env).zone_create(realm_token, zonegroup_name,
-                    zone_name, endpoints, start_radosgw)
+            retval, out, err = RGWAM(self.env).zonegroup_create(realm_token,
```
The realm name should also be passed into this function. As I mention below a realm pull does not need to happen in this situation. It should be removed.

rkachach replied:
I removed this code as it was not used (and not well tested). We can have it back later if we really need this functions.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +668,158 @@ def realm_rm_zone_creds(self, realm_token_b64):
+        realm_token = RealmToken.from_base64_str(realm_token_b64)
+        access_key = realm_token.access_key
+        secret = realm_token.secret
+        try:
```
This realm pull should be removed. It is unnecessary and might overwrite uncommitted multisite state. If you need info on multisite state you can always run `radosgw-admin period get`, but I think in this case you can just pass in the realm name or have it be part of the realm_token.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -193,46 +190,54 @@ def __init__(self, env: EnvArgs):
-
-        return RGWAdminJSONCmd(ze).run(params)
+        try:
+            ze = ZoneEnv(self.env)
```
Knit: why is this type called `ZoneEnv` not `MultisiteEnv`? I feel like that would be a more suitable name given how this type is used in all calls of `RGWAdminJSONCmd`

rkachach replied:
This is legacy code, I can rename the variable as part of other PR to not introduce a lot of noise on this one.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -193,46 +190,54 @@ def __init__(self, env: EnvArgs):
         return RGWAdminJSONCmd(ze).run(params)
 
-    def pull(self, url, access_key, secret, set_default=False):
+    def pull(self, realm, url, access_key, secret, set_default=False):
```
What is `set_default` being used for? I think you could remove it.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+            logging.info('Created system user: %s' % sys_user.uid)
+            access_key = sys_user.keys[0].access_key if sys_user and sys_user.keys else ''
+            secret_key = sys_user.keys[0].secret_key if sys_user and sys_user.keys else ''
+            sys_user.add_key(access_key, secret_key)
```
Out of curiosity what's the point of `add_key` here?

rkachach replied:
Good catch! I think it's useless and the previous lines as well (they are there because I tried to keep the original code logic as it's). I adjusted/fixed the code in the last commit.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+        secret = sys_user.keys[0].secret_key
+        self.zone_op().modify(zone, zonegroup, None, None, access_key, secret)
+        self.update_period(realm, zonegroup)
+        self.create_normal_user(zonegroup, zone)
```
Why is a normal user being created? IMO the multisite orchestration code should not be creating any users other than what is needed for multisite.

rkachach replied:
I removed the normal user creation as in fact it's not needed for RGW multisite ops. If the user do need one then he has to create it manually.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+            rgw_spec.update_endpoints = True
+            self.env.mgr.apply_rgw(rgw_spec)
 
     def realm_new_zone_creds(self, realm_name, endpoints, sys_uid):
```
why is a system user getting created here? If doing a `radosgw-admin user info --uid ${sys_uid}` doesn't give you back the keys for sys_uid you should throw an error.

rkachach replied:
I have no idea in this case since this code is legacy and I didn't change this function.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -541,40 +609,28 @@ def realm_new_zone_creds(self, realm_name, endpoints, sys_uid):
+        access_key = realm_token.access_key
         try:
             user_info = self.user_op().info(master_zone, master_zg, access_key=access_key)
         except RGWAMException as e:
```
This exception says it is failing to create a system user when it's doing a `user info` command not a `user create`.

rkachach replied:
Fixed

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +668,158 @@ def realm_rm_zone_creds(self, realm_token_b64):
-            realm_info = self.realm_op().pull(
-                realm_token['endpoint'], access_key, secret, set_default=True)
+            # We only pull the realm if we are on a secondary cluster
+            if not realm_token.is_primary and realm_token.endpoint is not None:
```
This realm pull is not necessary. The realm pull only needs to happen when you are creating the secondary zone for the first time, not modifying it.

rkachach replied:
AFAIR from my testing without this pull the modify zone fails (when updating the end-points of the secondary zone). I have to double check what went wrong on those cases.

alimaredia replied:
This seems strange and shouldn't be the case, let me know how double checking goes. We can do this together if you'd like.

rkachach replied:
@alimaredia I double checked this and find out that indeed it was not needed so I just went and removed the corresponding code. Not sure why I got errors in the past when not pulling the zone during the modify.

alimaredia commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +668,158 @@ def realm_rm_zone_creds(self, realm_token_b64):
+        logging.debug(period.to_json())
+
+        if start_radosgw:
+            secondary_realm_token = RealmToken(realm_name,
```
What's the point of the `secondary_realm_token`? You won't be pulling from this zone if you were setting up another cluster.

rkachach replied:
@alimaredia it's used to call back the rgw module from cephadm (to update the end-points once the daemons are created). That's the only purpose of it.

alimaredia replied:
By update the endpoints do you mean update the endpoints for the token or to update the endpoints of the rgw daemons that are stored internally somehow? 

The Ceph cluster admin deploying multisite would never use these secondary_realm_tokens, so can we get rid of them to avoid confusion?

rkachach replied:
@alimaredia the token in this case is used `internally` (by cephadm) to update the endpoints of the rgw secondary zone by using the info from the cephadm daemons. 

Anyway, in general the tokens are not stored: they are generated on the fly by using the current real information. So if a user runs `ceph rgw realm tokens` they should get up-to-date tokens for each realm.

alimaredia replied:
Just to be clear the up-to-date tokens for each realm have endpoints in the master zone right?

rkachach replied:
> Just to be clear the up-to-date tokens for each realm have endpoints in the master zone right?

Yes, we always look for the master zone of the realm and use its endpoints when generating the tokens.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +191,84 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+    @CLICommand('rgw zone update', perm='rw')
+    def update_zone_info(self, realm_name: str, zonegroup_name: str, zone_name: str, realm_token: str, endpoints: List[str]):
+        try:
+            retval, out, err = RGWAM(self.env).zone_modify(realm_name,
```
I'm not sure where you'll add this but if you're setting `--endpoints` for a master zone, it needs to also be set for the master zonegroup via `radosgw-admin zonegroup modify --endpoints {endpoints list}`.

alimaredia replied:
@rkachach did you see this comment? This was one of the few steps that was missed in this PR

rkachach replied:
@alimaredia just to understand this a little bit, what endpoints should be set in the master zonegroup? should these be the ones from the master zone only or we ` should use ALL` the endpoints (from master and secondary zones)?

alimaredia replied:
just the ones from the **master zone**, not ALL endpoints from all zones.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -166,10 +277,10 @@ def _cmd_rgw_zone_create(self,
+                                 zonegroup_name: Optional[str] = None,
+                                 zone_name: Optional[str] = None,
+                                 update: Optional[bool] = False):
         """Bootstrap new rgw zone that syncs with existing zone"""
```
What does realm_reconcile do? Does it update zone's endpoints?

rkachach replied:
This is legacy code. I'm planning to remove these cmds from the docs as no body are using them nor they are really well tested to have them documented (so some user can go and user them). WDYT?

alimaredia replied:
If you don't see any use for it, I think it's fair game to remove this code.

rkachach replied:
By now I'll just remove the cmds from the docs keeping the code so users don't make use of these commands.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -134,8 +180,7 @@ def _cmd_rgw_realm_new_zone_creds(self,
     @CLICommand('rgw realm zone-creds remove', perm='rw')
-    def _cmd_rgw_realm_rm_zone_creds(self,
-                                 realm_token : Optional[str] = None):
+    def _cmd_rgw_realm_rm_zone_creds(self, realm_token: Optional[str] = None):
```
what's the use case for removing the zone credentials?

rkachach replied:
Same as previous. This is legacy non tested code, I'd suggest removing these commands at least from the docs. We can include them back once they are well tested.

alimaredia commented:
@rkachach how are endpoints getting set in the zone? Can the user provide the endpoints list in one of the specs before running `ceph rgw zone update`? This should be the preferred if not only way to set them.

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+
+.. prompt:: bash #
+
+  ceph rgw realm tokens | jq
```
In fact this is just for a pretty print (we are not filtering for any specific entry). I added the `jq` to make the output clear as the tokens normally are very long.

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+
+User can use the token to create and synchronize a secondary zones
+on another cluster with the master zone by using `ceph rgw zone create`
+command and proving the corresponding token.
```
Fixed

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+
+.. note:: The spec file used by RGW has the same format as the one used by cephadm. Thus,
+          the user can provide any cephadm rgw supported parameter as any other advanced
+          configuration items such as SSL certificates etc.
```
fixed

rkachach commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -814,6 +814,28 @@ def update_progress() -> None:
+                    rc, out, err = self.mgr.mon_command(zone_update_cmd)
+                    rgw_spec.update_endpoints = (rc != 0)  # keep trying on failure
+                    if rc != 0:
+                        self.log.error(f'Error when trying to update rgw zone {err}.. keep trying')
```
@adk3798 good idea, I changed the code to raise a health warning.

rkachach commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -814,6 +814,28 @@ def update_progress() -> None:
+                    rc, out, err = self.mgr.mon_command(zone_update_cmd)
+                    rgw_spec.update_endpoints = (rc != 0)  # keep trying on failure
+                    if rc != 0:
+                        self.log.error(f'Error when trying to update rgw zone {err}.. keep trying')
```
Fixed

rkachach commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -814,6 +814,28 @@ def update_progress() -> None:
 
+            if service_type == 'rgw':
+                rgw_spec = cast(RGWSpec, spec)
+                if rgw_spec.update_endpoints and rgw_spec.rgw_realm_token is not None:
+                    ep = []
+                    for s in self.mgr.cache.get_daemons_by_service(rgw_spec.service_name()):
+                        if s.ports:
+                            for p in s.ports:
+                                ep.append(f'http://{s.hostname}:{p}')
+                    zone_update_cmd = {
+                        'prefix': 'rgw zone update',
+                        'realm_name': rgw_spec.rgw_realm,
+                        'zonegroup_name': rgw_spec.rgw_zonegroup,
+                        'zone_name': rgw_spec.rgw_zone,
+                        'realm_token': rgw_spec.rgw_realm_token,
+                        'endpoints': ep,
+                    }
+                    self.log.debug(f'rgw cmd: {zone_update_cmd}')
+                    rc, out, err = self.mgr.mon_command(zone_update_cmd)
+                    rgw_spec.update_endpoints = (rc != 0)  # keep trying on failure
+                    if rc != 0:
+                        self.log.error(f'Error when trying to update rgw zone {err}.. keep trying')
```
@adk3798 code moved to its own func

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,30 +91,55 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
+                                     rgw_frontend_port=port,
+                                     placement=placement_spec)]
+            else:
+                return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr='Invalid arguments: -h or --help for usage')
```
Agree, fixed.

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,30 +91,55 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
         try:
-            retval, out, err = RGWAM(self.env).realm_bootstrap(realm_name, zonegroup_name,
-                    zone_name, endpoints, sys_uid, uid, start_radosgw)
+            if inbuf:
+                rgw_specs = self._parse_rgw_specs(inbuf)
+            elif (realm_name and zonegroup_name and zone_name):
+                placement_spec = PlacementSpec.from_string(placement) if placement else None
+                rgw_specs = [RGWSpec(rgw_realm=realm_name,
+                                     rgw_zonegroup=zonegroup_name,
+                                     rgw_zone=zone_name,
+                                     rgw_frontend_port=port,
+                                     placement=placement_spec)]
+            else:
+                return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr='Invalid arguments: -h or --help for usage')
```
Fixed

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,30 +91,55 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
+        specs = [o for o in yaml_objs if o is not None]
+        rgw_specs = []
+        for spec in specs:
+            # TODO(rkachach): should we use a new spec instead of RGWSpec here!
```
yes :+1:

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -99,30 +91,55 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
-        return HandleCommandResult(retval=retval, stdout=out, stderr=err)
+        return HandleCommandResult(retval=0, stdout="Realm(s) created correctly. Please, use 'ceph rgw realm tokens' to get the token.", stderr='')
+
+    def _parse_rgw_specs(self, inbuf: Optional[str] = None):
```
Fixed

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +162,83 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+    @CLICommand('rgw realm tokens', perm='r')
+    def list_realm_tokens(self):
+        realms_info = []
+        for realm_info in RGWAM(self.env).get_realms_info():
```
Good idea, I added a try-catch

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +162,83 @@ def _cmd_rgw_realm_rm_zone_creds(self,
+                             inbuf: Optional[str] = None):
         """Bootstrap new rgw zone that syncs with existing zone"""
+        try:
+            if inbuf:
+                rgw_specs = self._parse_rgw_specs(inbuf)
+            elif (zone_name and realm_token):
+                placement_spec = PlacementSpec.from_string(placement) if placement else None
+                rgw_specs = [RGWSpec(rgw_realm_token=realm_token,
+                                     rgw_zone=zone_name,
+                                     rgw_frontend_port=port,
+                                     placement=placement_spec)]
+            else:
+                return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr='Invalid arguments: -h or --help for usage')
```
Fixed

rkachach commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -886,6 +891,8 @@ def __init__(self,
 
         #: The RGW realm associated with this service. Needs to be manually created
         self.rgw_realm: Optional[str] = rgw_realm
+        #: The RGW zonegroup associated with this service. Needs to be manually created
```
Fixed

rkachach commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -193,46 +190,54 @@ def __init__(self, env: EnvArgs):
+            output = RGWAdminJSONCmd(ze).run(params)
+            return output.get('realms') or []
+        except RGWAMException:
+            # in case the realm list is empty an exception is raised
```
@adk3798 not sure if there other conditions but if no realms exist we get an exception and an error as the following:

```
[ceph: root@ceph-node-0 /]# radosgw-admin realm list
failed to list realms: (2) No such file or directory
[ceph: root@ceph-node-0 /]# echo $?
2
```
This is something that rgw does (IMHO we shouldn't fail to list realms when the list is empty). Anyway, I'll add a log with the full exception just in case.

rkachach commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -193,46 +190,54 @@ def __init__(self, env: EnvArgs):
+            output = RGWAdminJSONCmd(ze).run(params)
+            return output.get('zonegroups') or []
+        except RGWAMException:
+            return []
```
Fixed in the same way

rkachach commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -251,6 +256,15 @@ class ZoneOp:
+            output = RGWAdminJSONCmd(ze).run(params)
+            return output.get('zones') or []
+        except RGWAMException:
+            return []
```
same fix

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,36 +19,111 @@ RGW Realm Operations
+
+Following is an example of zone spec file:
+
+.. code-block:: yaml
+
+  rgw_realm: myrealm
+  rgw_zonegroup: myzonegroup
+  rgw_zone: my-secondary-zone
+  rgw_realm_token: <token>
+  placement:
+    hosts:
+     - ceph-node-1
+     - ceph-node-2
+  spec:
+    rgw_frontend_port: 5500
```
@adk3798 in fact for the secondary zone creation we only need `rgw_zone` and `rgw_realm_token`. I lef the other fields because the RGW spec was checking for them. I have updated the code to only check for the token/zone in case of zone creation.

rkachach commented:
> @rkachach how are endpoints getting set in the zone? Can the user provide the endpoints list in one of the specs before running `ceph rgw zone update`? This should be the preferred if not only way to set them.

One of the nice things about this whole PR is that it requieres zero intervention from the user. He just needs to launch the bootstrap command providing the spec (or the arguments) and everything is done: realms/zonegroup/zones are created, rgw daemons are launched and the end-points are updated automatically. Since cephadm controls the endpoints I don't see the user case where the user has to provide the endpoints manually. Please, can you provide some example where this manual intervention from the user would be needed?

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +433,122 @@ def zone_op(self):
+                                                  uid_prefix='user-sys',
                                                   is_system=True)
+            sys_user = RGWUser(sys_user_info)
+            logging.info(f'Created system user: {sys_user.uid} on {realm.name}/{zonegroup.name}/{zone.name}')
```
```
ceph/rgw/rgwam_core.py:489:101: E501 line too long (109 > 100 characters)
ceph/rgw/rgwam_core.py:498:101: E501 line too long (104 > 100 characters)
```

rkachach replied:
fixed

adk3798 commented on src/python-common/ceph/rgw/types.py:
```diff
@@ -47,13 +50,23 @@ def to_json(self):
+            realm_token = json.loads(realm_token_s)
+            return cls(**realm_token)
+        except binascii.Error:
+            return None
```
```
ceph/rgw/types.py:71:1: E302 expected 2 blank lines, found 1
```

rkachach replied:
fixed

adk3798 reviewed (commented):
looks like all of my comments have been addressed pending any potential continuing discussions on the endpoints. Will do another in depth look over around when we start testing this but it seems to generally look good.

alimaredia commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -146,18 +162,83 @@ def _cmd_rgw_realm_rm_zone_creds(self,
     @CLICommand('rgw zone create', perm='rw')
     def _cmd_rgw_zone_create(self,
-                             realm_token : Optional[str] = None,
-                             zonegroup_name: Optional[str] = None,
```
@adk3798 @rkachach the name of the zone group should be in the `radosgw-admin zone create` command. A new zone group will not need to be created though. Doing a successful `realm pull` will bring the zone group over from the primary cluster along with the realm.

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -246,11 +253,26 @@ def create(self, realm: EntityKey, zg: EntityKey = None, endpoints=None, is_mast
+        ze = ZoneEnv(self.env, realm=realm, zg=zg)
+        params = ['zonegroup', 'modify']
+        opt_arg(params, '--endpoints', endpoints)
+        return RGWAdminJSONCmd(ze).run(params)
```
```
ceph/rgw/rgwam_core.py:262:1: E302 expected 2 blank lines, found 1
```
this stuff can be found locally by running tox in src/python-common btw

rkachach replied:
Fixed, thanks for the tip :+1:

rkachach commented:
jenkins test make check

rkachach commented:
jenkins test make check arm64

alimaredia commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -871,6 +874,9 @@ def __init__(self,
                  custom_configs: Optional[List[CustomConfig]] = None,
+                 rgw_realm_token: Optional[str] = None,
+                 update_endpoints: Optional[bool] = False,
+                 endpoints: Optional[str] = None  # commad separated endpoints list
```
could we call this `zone_endpoints` just so that this list of endpoints is not confused with others.

rkachach replied:
I renamed the variable

alimaredia commented on doc/mgr/rgw.rst:
```diff
@@ -18,57 +19,110 @@ RGW Realm Operations
+
+Following is an example of RGW mutlisite spec file:
+
+.. code-block:: yaml
```
In the example specs can you have one with endpoints in there, with an explanation of what their purpose is? If you start this section I can suggest some text.

rkachach replied:
Sure, I added a new example with endpoints

rkachach commented:
jenkins test make check

rkachach commented:
jenkins test make check arm64

rkachach commented:
jenkins test windows

rkachach commented:
jenkins test make check arm64

rkachach commented:
jenkins test dashboard cephadm

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -386,101 +439,124 @@ def zone_op(self):
+        if realm_name in self.realm_op().list():
+            raise RGWAMException(f'Realm {realm_name} already exists')
+        if zonegroup_name in self.zonegroup_op().list():
+            raise RGWAMException(f'ZonegroupOp {zonegroup_name} already exists')
```
```suggestion
            raise RGWAMException(f'Zonegroup {zonegroup_name} already exists')
```
also, will this list only contain zonegroups from within the user provided realm? As in, if there was another realm with the same zonegroup name would this fail and if so, should it?

rkachach replied:
AFAIK the zonegroup/zones must be unique across all the realms but maybe @alimaredia can confirm.

adk3798 commented on src/python-common/ceph/rgw/rgwam_core.py:
```diff
@@ -612,99 +676,128 @@ def realm_rm_zone_creds(self, realm_token_b64):
+        realm = EntityID(realm_id)
+        period_info = self.period_op().get(realm)
+        period = RGWPeriod(period_info)
+        logging.info('Period: ' + period.id)
```
should this be logged at debug level? Would we need this period id for anything?

rkachach replied:
Changed to debug. I guess it may be useful for debugging. It was there in the original code so I left the log message.

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,57 +19,120 @@ RGW Realm Operations
+          configuration features such as SSL certificates etc.
+
+Users can also specify custom zone endpoints in the spec (or through the cmd line). In this case, no
+cephadm daemons will be launched. Following is an example RGW spec file with zone endpoints:
```
> to clarify, this is saying if certain fields are added in the spec, it won't pass the spec to the orchestrator for daemon deployment?

The answer is yes, because in case the user specifies custom endpoints we cannot force cephadm to use them and IMHO it doesn't make sense to launch cephadm daemons in this case because we won't be able to listen in the specified endpoints. WDYT?

adk3798 replied:
eh, I think it's a bit of a confusing system that some certain field being present in the spec decides whether we'll tell the orchestrator to deploy daemons, but I'm okay with it for now and maybe we can come up with some more explicit way of handling it later.

rkachach replied:
Yeah, I agree. Probably we have to show some more explicit warning message in this case as rgw daemons are not launched by cephadm. I  expect this something that a user would use only if they are deploying the daemons separately themselves. Anyway this is something that we could address later :+1:

rkachach commented on doc/mgr/rgw.rst:
```diff
@@ -18,57 +19,120 @@ RGW Realm Operations
+.. code-block:: yaml
+
+  rgw_zone: my-secondary-zone
+  rgw_realm_token: <token>
```
> is this token field only useful for secondary clusters where we want to sync with a realm on the primary cluster?

Yes, actually is only used when creating a secondary zone

rkachach commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -561,6 +561,35 @@ def _apply_service_config(self, spec: ServiceSpec) -> None:
+        if rc != 0:
+            self.log.error(f'Error when trying to update rgw zone: {err}')
+            self.mgr.set_health_warning('CEPHADM_RGW', 'Cannot update rgw endpoints', 1,
+                                        [f'Cannot update rgw endpoints for daemon {rgw_spec.service_name()}'])
```
Fixed

rkachach commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -85,6 +98,18 @@ def config_notify(self) -> None:
+                if available:
+                    return func(self, *args, **kwargs)
+                else:
+                    err_msg = f"Cephadm is not available. Please enable cephadm by 'ceph mgr module enable cephadm'."
```
Fixed

adk3798 commented:
https://pulpito.ceph.com/adking-2022-12-02_21:56:44-orch:cephadm-wip-adk-testing-2022-12-02-0932-distro-default-smithi/

Failures tracked by:

- https://tracker.ceph.com/issues/49287
- https://tracker.ceph.com/issues/57755
- https://tracker.ceph.com/issues/57771
- https://tracker.ceph.com/issues/58145
- https://tracker.ceph.com/issues/58173

It's a lot of failures, but other than the the last one, they are all things we were aware of before the run (even the dead mgr-nfs-upgrade job had a `mount.nfs: mounting smithi174:/fake failed, reason given by server: No such file or directory` issue). The last one (https://tracker.ceph.com/issues/58173) as well seems to be a failure on a test setting flags on a pool, and given all the PRs in this run strictly touched mgr modules, I think it's unlikely any of the attached PRs caused the failures. Overall, I think the failures should only block merging on PRs related to nfs due to https://tracker.ceph.com/issues/58145 making it so we can't properly run our nfs related tests currently.

adk3798 commented:
jenkins retest this please

adk3798 commented:
jenkins test make check

rkachach commented:
jenkins test dashboard cephadm
